File: backend/app/models/meeting.py
"""
Pydantic schemas for Meeting Intelligence (DailyMemory).
"""

from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field


# --- Meeting ---

class MeetingCreate(BaseModel):
    title: str
    start_time: datetime
    end_time: Optional[datetime] = None
    participants: Optional[list[str]] = None
    calendar_event_id: Optional[str] = None


class MeetingUpdate(BaseModel):
    title: Optional[str] = None
    end_time: Optional[datetime] = None
    participants: Optional[list[str]] = None
    status: Optional[Literal["scheduled", "recording", "processing", "completed", "failed"]] = None


class MeetingResponse(BaseModel):
    id: str
    title: str
    calendar_event_id: Optional[str]
    start_time: datetime
    end_time: Optional[datetime]
    duration_seconds: Optional[int]
    participants: Optional[list[str]]
    status: str
    created_at: datetime

    model_config = {"from_attributes": True}


class MeetingListResponse(BaseModel):
    meetings: list[MeetingResponse]
    total: int


# --- Recording ---

class RecordingStartRequest(BaseModel):
    meeting_id: Optional[str] = None
    title: Optional[str] = None
    source: Literal["system", "mic", "mixed"] = "mixed"


class RecordingStatusResponse(BaseModel):
    is_recording: bool
    meeting_id: Optional[str] = None
    duration_seconds: float = 0
    audio_levels: Optional[dict] = None


class RecordingMetadataResponse(BaseModel):
    meeting_id: str
    duration_seconds: Optional[float]
    file_size_bytes: Optional[int]
    format: str
    sample_rate: int
    channels: int

    model_config = {"from_attributes": True}


# --- Transcript ---

class TranscriptSegmentResponse(BaseModel):
    id: int
    speaker: Optional[str]
    start_time: float
    end_time: float
    text: str
    confidence: Optional[float]

    model_config = {"from_attributes": True}


class TranscriptResponse(BaseModel):
    meeting_id: str
    segments: list[TranscriptSegmentResponse]
    total_segments: int


# --- Summary ---

class SummaryResponse(BaseModel):
    id: str
    meeting_id: str
    summary_text: str
    key_topics: Optional[list[str]]
    action_items: Optional[list[dict]]
    decisions: Optional[list[str]]
    model_used: Optional[str]
    created_at: datetime

    model_config = {"from_attributes": True}


# --- Chat ---

class MeetingChatRequest(BaseModel):
    message: str = Field(..., max_length=5000)
    meeting_id: Optional[str] = None


class MeetingChatSource(BaseModel):
    meeting_id: str
    meeting_title: str
    text: str
    speaker: Optional[str]
    timestamp: Optional[float]


class MeetingChatResponse(BaseModel):
    answer: str
    sources: list[MeetingChatSource]


# --- Search ---

class MeetingSearchResult(BaseModel):
    meeting_id: str
    meeting_title: str
    snippet: str
    score: float
    timestamp: Optional[float]
    speaker: Optional[str]


class MeetingSearchResponse(BaseModel):
    results: list[MeetingSearchResult]
    total: int
    query: str


# --- Speaker Mapping ---

class SpeakerMappingResponse(BaseModel):
    id: int
    meeting_id: str
    speaker_label: str
    display_name: str

    model_config = {"from_attributes": True}


class SpeakerMappingUpdate(BaseModel):
    speaker_label: str
    display_name: str

File: frontend/src/types/meeting.ts
// Meeting Intelligence Types

export type MeetingStatus = 'scheduled' | 'recording' | 'processing' | 'completed' | 'failed'

export interface Meeting {
  id: string
  title: string
  calendar_event_id: string | null
  start_time: string
  end_time: string | null
  duration_seconds: number | null
  participants: string[] | null
  status: MeetingStatus
  created_at: string
}

export interface MeetingListResponse {
  meetings: Meeting[]
  total: number
}

export interface MeetingCreate {
  title: string
  start_time: string
  end_time?: string
  participants?: string[]
  calendar_event_id?: string
}

export interface TranscriptSegment {
  id: number
  speaker: string | null
  start_time: number
  end_time: number
  text: string
  confidence: number | null
}

export interface TranscriptResponse {
  meeting_id: string
  segments: TranscriptSegment[]
  total_segments: number
}

export interface MeetingSummary {
  id: string
  meeting_id: string
  summary_text: string
  key_topics: string[] | null
  action_items: ActionItem[] | null
  decisions: string[] | null
  model_used: string | null
  created_at: string
}

export interface ActionItem {
  owner: string
  description: string
  due: string | null
}

export interface RecordingStatus {
  is_recording: boolean
  meeting_id: string | null
  duration_seconds: number
  audio_levels: AudioLevels | null
}

export interface AudioLevels {
  system: number
  mic: number
  mixed: number
}

export interface RecordingMetadata {
  meeting_id: string
  duration_seconds: number | null
  file_size_bytes: number | null
  format: string
  sample_rate: number
  channels: number
}

export interface MeetingChatMessage {
  role: 'user' | 'assistant'
  content: string
  sources?: MeetingChatSource[]
}

export interface MeetingChatSource {
  meeting_id: string
  meeting_title: string
  text: string
  speaker: string | null
  timestamp: number | null
}

export interface MeetingSearchResult {
  meeting_id: string
  meeting_title: string
  snippet: string
  score: number
  timestamp: number | null
  speaker: string | null
}

export interface MeetingSearchResponse {
  results: MeetingSearchResult[]
  total: number
  query: string
}

export interface SpeakerMapping {
  id: number
  meeting_id: string
  speaker_label: string
  display_name: string
}

export interface LiveTranscriptSegment {
  type: 'segment'
  id: number
  start: number
  end: number
  text: string
}
